host.py

```diff
@@ -296,20 +296,43 @@ def copy_to(self, src_file: str, dst_file: str) -> None:
     def need_sudo(self) -> None:
         self.sudo_needed = True
 
-    def run(self, cmd: str, log_level: int = logging.DEBUG, env: Dict[str, str] = os.environ.copy()) -> Result:
+    def run(
+        self,
+        cmd: str,
+        log_level: int = logging.DEBUG,
+        env: Optional[Dict[str, str]] = None,
+        *,
+        env_extra: Optional[Dict[str, str]] = None,
+    ) -> Result:
         if self.sudo_needed:
             cmd = "sudo " + cmd
 
         logger.log(log_level, f"running command {cmd} on {self._hostname}")
         if self.is_localhost():
-            ret_val = self._run_local(cmd, env)
+            ret_val = self._run_local(cmd, env, env_extra=env_extra)
         else:
-            ret_val = self._run_remote(cmd, log_level)
+            if env is not None:
+                raise ValueError("remote host does not support full environment. Use env_extra instead")
```

Comment:
I think the link is wrong. Or I just don't follow. What exactly do you mean?

Reply:
That is cumbersome (to have extra lines to copy the environment, which we basically *always* want).

In any case, the problem is that with the `env` parameter, the function does not know which variables the user intentionally wants to set/unset and which are just copied. Hence, there are backends that don't support resetting the entire environment (`sudo_needed`, `_run_remote()`). The `env` argument thus only works for `_run_local()`.

Reply:
+1

Reply:
When you run under sudo or via SSH, then the environment variables in that environment are entirely different. It makes no sense to copy the current environment and try to apply it in another one. the only thing that makes sense w.r.t. environment variables, is to explicitly set/unset a variable. The `env_extra` argument supports that. The `env` argument not, because we cannot distinguish between variables that are intentionally set (not set) and which are just copied over.

Point in case:
- sudo does not have an option, to keep the callers environment as a whole. You can only set individual variables.
- `paramiko.SSHClient.exec_command()` has an argument `environment`, but that behaves as the `env_extra` argument (because the semantics of `env` make not sense) (*).

(*) also, usually setting environment variables that way is rejected by SSH anyway, which is why `_run_remote()` does not use the `environment` parameter.

Reply:
When I call `run`, I would copy the env and add what I need to change before passing that in. Are you saying that that doesn't work?

Reply:
I am saying that. It doesn't work with remote hosts (`Host._run_remote()`) and with sudo (`self.sudo_needed`). It only works with `Host._run_local()`. The `env` argument thus breaks the abstraction of `Host.run()`.

Note that it both doesn't work on current `main` and with the changes of this branch. Except, with this branch, you get an exception that you are doing something that doesn't work.

Also, I also don't think `env` makes sense in those cases, meaning it cannot be made to work. `env_extra` is there for that.

Reply:
sudo needed seems like another problem to solve separately iiuc?

Reply:
Also, what bug are you solving? iow, what part of the code is failing without this PR?

Reply:
> sudo needed seems like another problem to solve separately iiuc?

I don't understand what you mean.

The branch makes `Host.needs_sudo()` work as one would expect (that is, work with `cmd` being a shell script -- as the `cmd` argument to `Host.run()` *is* a shell script otherwise). And have the arguments to `Host.run()` (`cwd`, `env_extra`) work together with sudo.

> Also, what bug are you solving? iow, what part of the code is failing without this PR?

Mainly, getting rid of os.chdir() (by supporting a `cwd` argument). The working directory is process wide, changing that for all threads (in a multithreadded application) is bad. I would use that on another WIP branch, and using `os.chdir()` in new code shouldn't be done.

While at it, make the `cmd` argument consistent between `LocalHost` and `RemoteHost`. While at it, support a env-like argument (`env_extra`), that works with other implementations beside `Host._run_local()`.